File: semaphores/rw_lock.py
#! /usr/bin/env python

import threading
import time

THREADS = 4
MAX_COUNT = 10000000



class RW(threading.Thread):
    mutex = threading.Lock()
    gate = threading.Lock()
    readers = 0

    counter = 0
    def __init__(self, threadID):
        super(RW, self).__init__()
        self.threadID = threadID

    def reader_lock(self):
        RW.gate.acquire()
        RW.readers += 1
        if RW.readers == 1:
            RW.mutex.acquire()
        RW.gate.release()

    def reader_unlock(self):
        RW.gate.acquire()
        RW.readers -= 1
        if RW.readers == 0:
            RW.mutex.release()
        RW.gate.release()

    def writer_lock(self):
        RW.mutex.acquire()

    def writer_unlock(self):
        RW.mutex.release()

    def run(self):
        print("Thread {}".format(self.threadID))

        for i in range(MAX_COUNT/THREADS):
            if i % 10:
                self.reader_lock()
                c = RW.counter
                self.reader_unlock()
            else:
                self.writer_lock()
                RW.counter += 1
                self.writer_unlock()

def main():
    threads = []

    for i in range(THREADS):
        t = RW(i)
        threads.append(t)

    # Start all threads
    for t in threads:
        t.start()

    # Wait for all threads to complete
    for t in threads:
        t.join()

    print("Counter value: %d Expected: %d\n" % (RW.counter, int(MAX_COUNT/10)))


if __name__ == "__main__":
    main()

File: channels/rw_lock.go
/* Implements RW lock exclusion using messages */

package main

import (
    "fmt"
    "runtime"
)

const (
    PROCS      = 4
    MAX_COUNT  = 10000000
    GOROUTINES = 4
)

type Empty struct{}
type RWLock struct {
    writer  chan Empty
    readers chan int
}

func NewRWLock() RWLock {
    var lock RWLock
    lock.writer = make(chan Empty, 1)
    lock.readers = make(chan int, 1)
    lock.writer <- Empty{}
    lock.readers <- 0
    return lock
}

func (l RWLock) readerLock() {
    readers := <-l.readers
    readers++
    if readers == 1 {
        <-l.writer
    }
    l.readers <- readers
}

func (l RWLock) readerUnlock() {
    readers := <-l.readers
    readers--
    if readers == 0 {
        l.writer <- Empty{}
    }
    l.readers <- readers
}

func (l RWLock) writerLock() {
    <-l.writer
}

func (l RWLock) writerUnlock() {
    l.writer <- Empty{}
}

var counter = 0

func run(id, counts int, done chan Empty, rwlock RWLock) {
    for i := 0; i < counts; i++ {
        if i%10 > 0 {
            rwlock.readerLock()
            // c := counter
            rwlock.readerUnlock()
        } else {
            rwlock.writerLock()
            counter++
            rwlock.writerUnlock()
        }
    }
    fmt.Printf("End %d counter: %d\n", id, counter)
    done <- Empty{}
}

func main() {
    runtime.GOMAXPROCS(PROCS)
    done := make(chan Empty, 1)
    rwlock := NewRWLock()

    for i := 0; i < GOROUTINES; i++ {
        go run(i, MAX_COUNT/GOROUTINES, done, rwlock)
    }

    for i := 0; i < GOROUTINES; i++ {
        <-done
    }

    fmt.Printf("Counter value: %d Expected: %d\n", counter, MAX_COUNT/10)
}
